third_party/move/move-bytecode-verifier/src/signature_v2.rs

```diff
@@ -21,9 +22,271 @@ use std::{
+/// `PROGRAM_TOO_COMPLEX`. It's important to note that no direct gas charges are applied based on
+/// this meter. Therefore it is considered a pseudo metering mechanism.
+struct Meter<'a> {
+    resolver: BinaryIndexedView<'a>,
+
+    cached_signature_sizes: RefCell<BTreeMap<SignatureIndex, usize>>,
+    balance: RefCell<u64>,
+}
+
+/// Calculates the allowed budget based on the size of a binary.
+///
+/// Formula: `base + per_byte * size`.
+fn calculate_budget(base: u64, per_byte: (u64, u64), size: usize) -> u64 {
+    base.saturating_add(u128::min(
+        (per_byte.0 as u128 * size as u128) / per_byte.1 as u128,
+        u64::MAX as u128,
+    ) as u64)
+}
+
+impl<'a> Meter<'a> {
+    fn charge(&self, amount: u64) -> PartialVMResult<()> {
+        let mut balance = self.balance.borrow_mut();
+        match balance.checked_sub(amount) {
+            Some(new_balance) => {
+                *balance = new_balance;
+                Ok(())
+            },
+            None => {
+                *balance = 0;
+                Err(PartialVMError::new(StatusCode::PROGRAM_TOO_COMPLEX)
+                    .with_message("program too complex for signature checker v2".to_string()))
+            },
+        }
+    }
+
+    fn meter_signature(&self, idx: SignatureIndex) -> PartialVMResult<()> {
+        let size = match self.cached_signature_sizes.borrow_mut().entry(idx) {
+            btree_map::Entry::Occupied(entry) => *entry.into_mut(),
+            btree_map::Entry::Vacant(entry) => {
+                let sig = self.resolver.signature_at(idx);
+                let size = sig
+                    .0
+                    .iter()
+                    .fold(0, |acc: usize, ty| acc.saturating_add(ty.num_nodes()));
+
+                *entry.insert(size)
+            },
+        };
+
+        self.charge(size as u64)?;
+
+        Ok(())
+    }
+
+    fn meter_signatures(&self) -> PartialVMResult<()> {
+        for sig_idx in 0..self.resolver.signatures().len() {
+            self.meter_signature(SignatureIndex(sig_idx as u16))?;
+        }
+        Ok(())
+    }
+
+    fn meter_function_instantiation(
+        &self,
+        func_inst_idx: FunctionInstantiationIndex,
+    ) -> PartialVMResult<()> {
+        let func_inst = self.resolver.function_instantiation_at(func_inst_idx);
+        self.meter_signature(func_inst.type_parameters)
+    }
+
+    fn meter_function_instantiations(&self) -> PartialVMResult<()> {
+        for func_inst_idx in 0..self.resolver.function_instantiations().len() {
+            self.meter_function_instantiation(FunctionInstantiationIndex(func_inst_idx as u16))?;
+        }
+        Ok(())
+    }
+
+    fn meter_struct_instantiation(
+        &self,
+        struct_inst_idx: StructDefInstantiationIndex,
+    ) -> PartialVMResult<()> {
+        let struct_inst = self
+            .resolver
+            .struct_instantiation_at(struct_inst_idx)
+            .expect("struct instantiation should exist");
```

Comment:
expects are problematic in the case someone can construct a variant that would fail here. Would really rather we emit an error that this is not a valid module.